src/state/wifi_collection.cpp

```diff
@@ -0,0 +1,102 @@
+#include "wifi_collection.h"
+#include "qjsonarray.h"
+
+
+namespace Valeronoi::state {
+
+wifi_collection::wifi_collection()
+{
+    clear();
+}
+
+void wifi_collection::clear()
+{
+    m_known_wifis.clear();
+    m_known_wifi_ids.clear();
+}
+
+int wifi_collection::get_or_create_wifi_id(Valeronoi::robot::WifiInformation wifi_info)
+{
+    int nRet = get_wifi_id(wifi_info.bssid());
+
+    if (nRet < 0) {
+        nRet = add_wifi(wifi_info);
+    }
+
+    return nRet;
+}
+
+int wifi_collection::add_wifi(Valeronoi::robot::WifiInformation wifi_info)
+{
+    m_known_wifis.append(wifi_info);
+    m_known_wifi_ids.append(m_known_wifis.size()-1);
+
+    if (m_known_wifis.size() != m_known_wifi_ids.size())
+    {
+        // out of sync.
+        throw "Vector size missmatch! this is bad";
+    }
+
+    emit signal_wifi_list_updated();
+    emit signal_new_wifi_added(wifi_info);
+
+    return m_known_wifi_ids.last();
+}
+
+QVector<Valeronoi::robot::WifiInformation> wifi_collection::get_known_wifis() const
+{
+    return m_known_wifis;
+}
+
+QJsonArray wifi_collection::get_json() const
+{
+    QJsonArray jArr;
+
+    for (int index = 0; m_known_wifis.size() > index ; ++index) {
+        QJsonObject wifi_access_point = m_known_wifis.at(index).get_json();
+        wifi_access_point.insert("index",m_known_wifi_ids.at(index));
+        jArr.append(wifi_access_point);
+    }
+
+    return jArr;
+}
+
+void wifi_collection::set_json(const QJsonArray &json)
+{
+    clear();
+
+    foreach (auto wifi_access_point_value, json) {
+        if (!wifi_access_point_value.isObject()) {
+            // Wrong format.
+        }
+        QJsonObject wifi_access_point = wifi_access_point_value.toObject();
+        int index = wifi_access_point["index"].toInt(-1);
+        if (index < 0) {
+            // error index.
+            // continue Mapping this as -1.
```

Comment:
?

Reply:
I rewrote the comment and renamed the index to wifi_id.....  
This if-case should probably never happen and be impossible.  
You could only reach it if the loaded Json Value is incorrect or tampered, because the wifi_ids (formarly index) should always be >0.  

So this is a unknown Parsing error and because there is no way to fix this, i would just continue anyways with mapping it as wifi_id -1. So you would see the Wifi Info, but probably all measured Data is lost / unmapped from the Wifi. :/

I could not come up with real soulution for this...  
Probably only a popUp for "currupted Data" could be a soulution.

Reply:
if the same parsing error happens in the wifi_id of the Mesaurements, they will be connected by the "unknown_wifi_id".  
This could scramble up measurements frome multiple WiFis, but your data would have been currupted anyways to come to this point.
So maybe it is the best approach to reastore something.

Reply:
I just tested it, it will screw up a lot of things here, so I will just throw an expection if this happens. Which will probably just close the Application, but i have no better solution for tihs.